src/compiler/checker.ts

```diff
@@ -10494,17 +10512,24 @@ namespace ts {
         }
 
         function getContextualTypeForJsxAttribute(attribute: JsxAttribute | JsxSpreadAttribute) {
+            // When we trying to resolve JsxOpeningLikeElement as a stateless function element, we will already give JSXAttributes a contextual type
+            // which is a type of the parameter  of the signature we are trying out. This is not the case if it is a statefull Jsx (i.e ReactComponenet class)
+            // So if that is the case, just return the type of the JsxAttribute in such contextual type with out going into resolving of the JsxOpeningLikeElement again
+            if ((<JsxAttributes>attribute.parent).contextualType) {

```

Comment:
This is probably wrong. Any time I played games with contextual typing, it turned out that I was missing the root of the problem. I need to read the rest of the review to see whether I can figure out what the root actually is, though.